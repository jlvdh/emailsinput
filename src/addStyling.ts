export default (
  element: HTMLElement,
  styles: { [index: string]: string }
): HTMLElement => {
  for (const key in styles) {
    (element as any).style[key] = styles[key];
  }
  return element;
};
