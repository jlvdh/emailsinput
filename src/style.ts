export const containerStyles = {
  border: '1px solid #C3C2CF',
  borderRadius: '4px',
  background: '#FFFFFF',
  padding: '6px 7px 8px 7px',
  height: '100%',
  overflowY: 'scroll',
};

export const emailLabelStyle = {
  display: 'inline-block',
  background: 'rgba(102, 153, 255, 0.2)',
  borderRadius: '100px',
  fontSize: '14px',
  lineHeight: '24px',
  color: '#050038',
  padding: '4px 10px',
  margin: '0px 4px 0 0',
};

export const emailErrorLabelStyle = {
  backgroundColor: 'white',
  borderBottom: '1px dashed #D92929',
  borderRadius: '0px',
  padding: '4px 0px',
  margin: '4px 8px 0px 4px',
};

export const emailDeleteButtonStyle = {
  backgroundImage: `url("data:image/svg+xml,%3C%3Fxml version='1.0' %3F%3E%3C!DOCTYPE svg PUBLIC '-//W3C//DTD SVG 1.1//EN' 'http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd'%3E%3Csvg enable-background='new 0 0 32 32' id='Слой_1' version='1.1' viewBox='0 0 32 32' xml:space='preserve' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink'%3E%3Cpath d='M17.459,16.014l8.239-8.194c0.395-0.391,0.395-1.024,0-1.414c-0.394-0.391-1.034-0.391-1.428,0 l-8.232,8.187L7.73,6.284c-0.394-0.395-1.034-0.395-1.428,0c-0.394,0.396-0.394,1.037,0,1.432l8.302,8.303l-8.332,8.286 c-0.394,0.391-0.394,1.024,0,1.414c0.394,0.391,1.034,0.391,1.428,0l8.325-8.279l8.275,8.276c0.394,0.395,1.034,0.395,1.428,0 c0.394-0.396,0.394-1.037,0-1.432L17.459,16.014z' fill='%23121313' id='Close'/%3E%3Cg/%3E%3Cg/%3E%3Cg/%3E%3Cg/%3E%3Cg/%3E%3Cg/%3E%3C/svg%3E")`,
  border: 'none',
  width: '17px',
  height: '17px',
  backgroundColor: 'transparent',
  backgroundSize: 'contain',
  backgroundRepeat: 'no-repeat',
  borderRadius: '0px',
  padding: '0px',
  position: 'relative',
  top: '5px',
  left: '3px',
};

export const emailInputStyle = {
  border: 'none',
  display: 'inline-block',
  outline: 'none',
  padding: '4px 0px',
  margin: '4px 8px 0px 4px',
};
