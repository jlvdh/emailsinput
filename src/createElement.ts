import addStyling from './addStyling';

const CreateElement = (
  type: string,
  options?: {
    attributes?: { [index: string]: string }
    styling?: { [index: string]: string}
  }
): HTMLElement => {
  const element = document.createElement(type);
  if (!options) return element;

  const {attributes, styling } = options;

  for (const key in attributes) {
    (element as any)[key] = attributes[key];
  }

  if(styling) {
    addStyling(element, styling);
  }
  
  return element;
};

export default CreateElement;
