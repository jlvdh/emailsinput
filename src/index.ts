import createElement from './createElement';
import isEmail from './isEmail';
import addStyling from './addStyling';
import {
  containerStyles,
  emailLabelStyle,
  emailErrorLabelStyle,
  emailDeleteButtonStyle,
  emailInputStyle,
} from './style';

class EmailsInput {
  private inputElement: HTMLInputElement;
  private className: string;

  constructor(targetElement: HTMLInputElement, className = '') {
    this.inputElement = targetElement;
    this.initialise();
    this.className = className;
  }

  private initialise = () => {
    this.inputElement.type = 'hidden';
    this.listenInputTargetChange();
    this.renderInitial();
  };

  private listenInputTargetChange = () => {
    const that = this;
    Object.defineProperty(this.inputElement, 'value', {
      get: function() {
        return this._value;
      },
      set: function(v) {
        this._value = v;
        that.renderLabels(this.previousSibling.firstChild);
      },
    });
  };

  private renderInitial = () => {
    // create initial HTML
    const container = createElement('div', {
      attributes: { className: this.className ?? 'emails-input' },
      styling: containerStyles,
    });
    const labelsContainer = createElement('span');
    const field = createElement('input', {
      attributes: { type: 'email', placeholder: 'add more people...' },
      styling: emailInputStyle,
    });

    container.appendChild(labelsContainer);
    container.appendChild(field);

    // listen for delete button
    labelsContainer.addEventListener('click', e => {
      if (
        (e.target as HTMLElement).classList.contains('__delete-email-button')
      ) {
        this.deleteEmail(+(e.target as HTMLElement).id.substring(15));
      }
    });

    /** focus on click in component */
    container.addEventListener('click', e => {
      const field = (e.currentTarget as HTMLElement).getElementsByTagName(
        'input'
      );
      field[0].focus();
    });

    field.addEventListener('keydown', this.handleKeyInput);
    field.addEventListener('paste', this.handlePaste);
    field.addEventListener('blur', this.handleBlur);

    this.inputElement.parentNode!.insertBefore(container, this.inputElement);
  };

  private getEmails = () => {
    if (!this.inputElement.value) return [];
    return this.inputElement.value.split(',');
  };

  private deleteEmail = (index: number) => {
    const emails = this.getEmails();
    emails.splice(index, 1);
    this.inputElement.value = emails.toString();
  };

  private addEmail = (email: string) => {
    if (!email) return;
    if (this.inputElement.value) {
      this.inputElement.value = `${this.inputElement.value}, ${email.trim()}`;
    } else {
      this.inputElement.value = email.trim();
    }
  };

  private renderLabels = (labelsContainer: HTMLElement) => {
    labelsContainer.innerHTML = '';
    const emails = this.getEmails();
    emails.forEach((email: string, index: number) => {
      const deleteButton = this.createDeleteButton(index);
      const emailLabel = this.createEmailLabel(email);
      emailLabel.appendChild(deleteButton);
      labelsContainer.appendChild(emailLabel);
    });
  };

  private createEmailLabel = (email: string) => {
    const emailLabel = createElement('div', { styling: emailLabelStyle });
    emailLabel.innerHTML = email;
    if (!isEmail(email)) {
      addStyling(emailLabel, emailErrorLabelStyle);
    }
    return emailLabel;
  };

  private createDeleteButton = (index: number) => {
    return createElement('button', {
      attributes: {
        id: `__delete-label-${index.toString()}`,
        className: '__delete-email-button',
      },
      styling: emailDeleteButtonStyle,
    });
  };

  private handleKeyInput = (e: KeyboardEvent) => {
    const { value } = e.target as HTMLInputElement;
    if (isActionKey(e.key)) {
      // prevent default to prevent ',' being added to input
      e.preventDefault();
      this.addEmail(value);
      (e.target as HTMLInputElement).value = '';
    }
  };

  private handlePaste = (e: ClipboardEvent) => {
    // using setTimeout as an alternative to clipboard data as it's not supported by IE11
    setTimeout(() => {
      let input = (e.target as HTMLInputElement).value.trim();
      if (input.charAt(input.length - 1) === ',') {
        input = input.slice(0, -1);
      }
      this.addEmail(input);
      (e.target as HTMLInputElement).value = '';
    }, 100);
  };

  private handleBlur = (e: FocusEvent) => {
    const { value } = e.target as HTMLInputElement;
    this.addEmail(value);
    (e.target as HTMLInputElement).value = '';
  };
}

const isActionKey = (code: string): boolean => {
  // check event.key
  switch (code) {
    case 'Enter':
    case ',':
      return true;
    default:
      return false;
  }
};

export const create = (div: HTMLInputElement) => {
  new EmailsInput(div);
};
