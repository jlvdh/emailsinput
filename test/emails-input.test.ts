import { create } from '../src';
import createElement from '../src/createElement';

const input = createElement('input', {
  attributes: { type: 'email', id: 'emailsinput' },
});
const container = createElement('div');
container.appendChild(input);
document.body.appendChild(container);

describe('emailsinput', () => {
  it('matches snapshot', () => {
    const input = document.getElementById('emailsinput');
    create(input);
    expect(container).toMatchSnapshot();
  });

  it('matches snapshot when one email is entered', () => {
    // const input = document.querySelector('input')
    (input as HTMLInputElement).value = 'jvdham@gmail.com';

    expect(container).toMatchSnapshot();
  });

  it('matches snapshot when multiple emails are entered', () => {
    // const input = document.querySelector('input')
    (input as HTMLInputElement).value =
      'jvdham@gmail.com,invalid.email,someother@mail.com';

    expect(container).toMatchSnapshot();
  });
});
