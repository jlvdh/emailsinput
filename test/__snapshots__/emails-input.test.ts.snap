// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emailsinput matches snapshot 1`] = `
<div>
  <div
    class="emails-input"
    style="border: 1px solid #c3c2cf; border-radius: 4px; background: rgb(255, 255, 255); padding: 6px 7px 8px 7px; height: 100%; overflow-y: scroll;"
  >
    <span />
    <input
      placeholder="add more people..."
      style="display: inline-block; outline: none; padding: 4px 0px; margin: 4px 8px 0px 4px;"
      type="email"
    />
  </div>
  <input
    id="emailsinput"
    type="hidden"
  />
</div>
`;

exports[`emailsinput matches snapshot when multiple emails are entered 1`] = `
<div>
  <div
    class="emails-input"
    style="border: 1px solid #c3c2cf; border-radius: 4px; background: rgb(255, 255, 255); padding: 6px 7px 8px 7px; height: 100%; overflow-y: scroll;"
  >
    <span>
      <div
        style="display: inline-block; background: rgba(102, 153, 255, 0.2); border-radius: 100px; font-size: 14px; line-height: 24px; color: rgb(5, 0, 56); padding: 4px 10px; margin: 0px 4px 0px 0px;"
      >
        jvdham@gmail.com
        <button
          class="__delete-email-button"
          id="__delete-label-0"
          style="width: 17px; height: 17px; background-color: transparent; background-size: contain; background-repeat: no-repeat; border-radius: 0px; padding: 0px; position: relative; top: 5px; left: 3px;"
        />
      </div>
      <div
        style="display: inline-block; background: rgba(102, 153, 255, 0.2); border-radius: 0px; font-size: 14px; line-height: 24px; color: rgb(5, 0, 56); padding: 4px 0px; margin: 4px 8px 0px 4px; background-color: white; border-bottom: 1px dashed #D92929;"
      >
        invalid.email
        <button
          class="__delete-email-button"
          id="__delete-label-1"
          style="width: 17px; height: 17px; background-color: transparent; background-size: contain; background-repeat: no-repeat; border-radius: 0px; padding: 0px; position: relative; top: 5px; left: 3px;"
        />
      </div>
      <div
        style="display: inline-block; background: rgba(102, 153, 255, 0.2); border-radius: 100px; font-size: 14px; line-height: 24px; color: rgb(5, 0, 56); padding: 4px 10px; margin: 0px 4px 0px 0px;"
      >
        someother@mail.com
        <button
          class="__delete-email-button"
          id="__delete-label-2"
          style="width: 17px; height: 17px; background-color: transparent; background-size: contain; background-repeat: no-repeat; border-radius: 0px; padding: 0px; position: relative; top: 5px; left: 3px;"
        />
      </div>
    </span>
    <input
      placeholder="add more people..."
      style="display: inline-block; outline: none; padding: 4px 0px; margin: 4px 8px 0px 4px;"
      type="email"
    />
  </div>
  <input
    id="emailsinput"
    type="hidden"
  />
</div>
`;

exports[`emailsinput matches snapshot when one email is entered 1`] = `
<div>
  <div
    class="emails-input"
    style="border: 1px solid #c3c2cf; border-radius: 4px; background: rgb(255, 255, 255); padding: 6px 7px 8px 7px; height: 100%; overflow-y: scroll;"
  >
    <span>
      <div
        style="display: inline-block; background: rgba(102, 153, 255, 0.2); border-radius: 100px; font-size: 14px; line-height: 24px; color: rgb(5, 0, 56); padding: 4px 10px; margin: 0px 4px 0px 0px;"
      >
        jvdham@gmail.com
        <button
          class="__delete-email-button"
          id="__delete-label-0"
          style="width: 17px; height: 17px; background-color: transparent; background-size: contain; background-repeat: no-repeat; border-radius: 0px; padding: 0px; position: relative; top: 5px; left: 3px;"
        />
      </div>
    </span>
    <input
      placeholder="add more people..."
      style="display: inline-block; outline: none; padding: 4px 0px; margin: 4px 8px 0px 4px;"
      type="email"
    />
  </div>
  <input
    id="emailsinput"
    type="hidden"
  />
</div>
`;
